plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go

```diff
@@ -87,24 +113,62 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
 				}
 			}
 			if matches {
-				counts[pod.Spec.NodeName]++
+				countsByNodeName[pod.Spec.NodeName]++
 				// Compute the maximum number of pods hosted on any node
-				if counts[pod.Spec.NodeName] > maxCount {
-					maxCount = counts[pod.Spec.NodeName]
+				if countsByNodeName[pod.Spec.NodeName] > maxCountByNodeName {
+					maxCountByNodeName = countsByNodeName[pod.Spec.NodeName]
 				}
 			}
 		}
 	}
 
+	maxCountByZone := 0
+	haveZones := false
+	countsByZone := map[string]int{}
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
+
+		count, found := countsByNodeName[node.Name]
+		if !found {
+			continue
+		}
+
+		zoneId := getZoneKey(node)
+		if zoneId == "" {
+			continue
+		}
+
+		haveZones = true
+		countsByZone[zoneId] += count
+		// Compute the maximum number of pods hosted in any zone
+		if countsByZone[zoneId] > maxCountByZone {
+			maxCountByZone = countsByZone[zoneId]
+		}
+	}
+
 	result := []schedulerapi.HostPriority{}
 	//score int - scale of 0-10
 	// 0 being the lowest priority and 10 being the highest
-	for _, node := range nodes.Items {
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
 		// initializing to the default/max node score of 10
 		fScore := float32(10)
-		if maxCount > 0 {
-			fScore = 10 * (float32(maxCount-counts[node.Name]) / float32(maxCount))
+		if maxCountByNodeName > 0 {
+			fScore = 10 * (float32(maxCountByNodeName-countsByNodeName[node.Name]) / float32(maxCountByNodeName))
+		}
+
+		// If there is zone information present, incorporate it
+		if haveZones {
+			zoneId := getZoneKey(node)
+			if zoneId != "" {
+				fScore += 20 * (float32(maxCountByZone-countsByZone[zoneId]) / float32(maxCountByZone))
```

Comment:
I have made it configurable, but left the TODO about "any way to justify 2/3 to zone-spreading".  I agree that it isn't entirely clear which is better.  I like your argument that losing a node is more likely than losing a zone.  Another strategy might be to first ensure that we have one pod in every zone; then we ensure that we have one pod on every node; then 2 in every zone; then 2 on every node etc.  But at the same time I don't think we want to make these hard constraints; if there's no capacity in a zone, then we should still schedule an RC even if they all end up in one AZ.  (The user can do hard-zone-spreading using node-selectors if needed).

I suspect the really tricky thing comes in figuring out how this interacts with our other scheduling preference functions...

My personal preference is to try it and see how we feel about the decisions it makes.  I don't think they are bad (particularly once we ignore deleted pods!), though they are not perfect.  To me the bigger problems are around the lack of rescheduling when nodes are created, and around the 'victim pod' not being priority aware as an RC is scaled down.